### PR TITLE
Align optimisation result

### LINKED ISSUE #271: Update optimisation class to return type `dict`
### Feature description

The optimisation class should return type dictionary including the following:

- The parameter set with the identified or optimised values (linking to #258)
- The optimisation variables (convergence for paramters and cost, optimisation type, number of iterations, etc.)
- If the forward model is pybamm based, the pybamm solution object.

### Motivation

A dictionary object allows for all the relevant information from the optimisation to be stored in one object. This makes it easier for debugging and post-processing the optimisation solution.

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Move the Result class into base_optimiser.py and use it to align the results returned by all optimisers.

## Issue reference
Fixes #271

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: pybop/optimisers/base_optimiser.py [hunk 1]
```diff
@@ -231,3 +232,26 @@ def set_allow_infeasible_solutions(self, allow=True):
...
+        The solution of the optimisation.
+    final_cost : float
+        The cost associated with the solution x.
+    nit : int
+        Number of iterations performed by the optimiser.
+    scipy_result : scipy.optimize.OptimizeResult, optional
+        The result obtained from a SciPy optimiser.
+    """
+
+    def __init__(self, x=None, final_cost=None, nit=None, scipy_result=None):
```
Comment (BradyPlanden) [inline pybop/optimisers/base_optimiser.py:253]: Let's add type hints here (where possible).

File: examples/standalone/optimiser.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import numpy as np
 from scipy.optimize import minimize
 
-from pybop import BaseOptimiser
+from pybop.optimisers.base_optimiser import BaseOptimiser, Result
```
Comment (BradyPlanden) [inline examples/standalone/optimiser.py:4]: I think we should add `Result` to `pybop/__init__.py` so the class is accessible throughout the library. This allows us to simply import it as `from pybop import Result` here and below.

File: pybop/optimisers/scipy_optimisers.py [hunk 3]
```diff
@@ -65,17 +65,19 @@ def _run(self):
...
-            The final cost associated with the best parameters.
+        result : pybop.Result
+            The result of the optimisation including the optimised parameter values and cost.
         """
-        self.result = self._run_optimiser()
+        result = self._run_optimiser()
 
-        self.result.final_cost = self.cost(self.result.x)
-        self._iterations = self.result.nit
+        self._iterations = result.nit
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:73]: I see that we set this attr, but don't use it the results class below. Is it needed for something else or do you think we can remove it? Alternatively, we could pass it directly to `Results` if that helps align the API between SciPy and Pints?

**Review by BradyPlanden — COMMENTED** [review]
Thanks @NicolaCourtier, this is a nice addition! It's a great step towards a larger `Results` class to be used across the library. A few comments to look at :)

File: pybop/optimisers/scipy_optimisers.py (hunk 3 above)
Comment (NicolaCourtier) [reply pybop/optimisers/scipy_optimisers.py:73]: This property is currently used in the tests, but we can switch to using the property of the result.

**Pull request merged**